src/oc_erchef/apps/chef_index/src/chef_index_batch.erl

```diff
@@ -111,6 +112,32 @@ init([]) ->
             MaxWait = envy:get(chef_index, search_batch_max_wait, 10, non_neg_integer),
             WrapperSize = wrapper_size(#chef_idx_batch_state{search_provider=SearchProvider}),
             CurrentSize = 0,
+
+
+            prometheus_gauge:declare([{name, chef_index_batch_current_batch_size_bytes},
+                                      {help, "The size (in bytes) of the batch currently being constructed"}]),
+            prometheus_gauge:declare([{name, chef_index_batch_current_batch_doc_count},
+                                      {help, "The number of documents in the batch currently being constructed"}]),
+            prometheus_gauge:declare([{name, chef_index_batch_inflight_flushes_count},
+                                      {help, "The number of inflight requests to the search index"}]),
+            prometheus_gauge:declare([{name, chef_index_batch_mailbox_size},
+                                      {help, "The number of items in the chef_index_batch processes mailbox"}]),
+
+            prometheus_counter:declare([{name, chef_index_batch_successful_docs_total},
+                                        {help, "The total number of documents thach chef_index_batch successfully processed"}]),
+            prometheus_counter:declare([{name, chef_index_batch_failed_docs_total},
+                                        {help, "The total number of documents that chef_index_batch failed to process"}]),
+
+            prometheus_histogram:declare([{name, chef_index_batch_queue_latency_ms},
+                                          {help, "The amount of time (in ms) that a document spent waiting in chef_index_batch before attempting to send it to the search index"},
+                                          {buckets,
+                                           [5, 10, 20, 50, 100, 250, 500, 750, 1000, 2000, 5000, 10000]}
```

Comment:
Noted, that probably the stddev isn't quite technically correct because the times are likely to a be skewed, poisson sort of distribution, but it still can be useful as a general guide to how spread the values are.

Reply:
> I'd also argue that anything with a histogram should have a total time counter; as they are complementary.

The histogram metrics also will emit a total count and a sum of all observations provided by the library.

Reply:
Here is some example data from this histogram:

```
# TYPE chef_index_batch_queue_latency_ms histogram
 # HELP chef_index_batch_queue_latency_ms The amount of time (in ms) that a document spent waiting in chef_index_batch before attempting to send it to the search index
chef_index_batch_queue_latency_ms_bucket{le="5"} 511
chef_index_batch_queue_latency_ms_bucket{le="10"} 964
chef_index_batch_queue_latency_ms_bucket{le="20"} 1020
chef_index_batch_queue_latency_ms_bucket{le="50"} 1028
chef_index_batch_queue_latency_ms_bucket{le="100"} 1031
chef_index_batch_queue_latency_ms_bucket{le="250"} 1031
chef_index_batch_queue_latency_ms_bucket{le="500"} 1031
chef_index_batch_queue_latency_ms_bucket{le="750"} 1031
chef_index_batch_queue_latency_ms_bucket{le="1000"} 1031
chef_index_batch_queue_latency_ms_bucket{le="2000"} 1031
chef_index_batch_queue_latency_ms_bucket{le="5000"} 1031
chef_index_batch_queue_latency_ms_bucket{le="10000"} 1031
chef_index_batch_queue_latency_ms_bucket{le="+Inf"} 1031
chef_index_batch_queue_latency_ms_count 1031
chef_index_batch_queue_latency_ms_sum 6114
```

Reply:
> We should be consistent on our bucket intervals across all of the histograms. I suggest that we cover from .1ms all the way up to 10000ms

What do you think about?

`[0.1,0.5,1,5,10,25,50,100,250,500,1000,2000,5000,10000]`

In principal I would like a lot more buckets, but the prometheus documentation has a lot of warnings about the performance implications of histograms, especially on "high cardinality" metrics.  Overall, I would prefer something like HdrHistogram.

Reply:
That looks great to me. I just want to make sure that they all use the same bins, so you can line two histograms up against each other and compare, and I want to make sure it's providing useful binning on systems with fast network connections. I hadn't known that there was a sum and count provided by histograms, that's enough to compute the average. My thinking is if we manage to totally miss the boat on the histogram binning, it would be nice to have a hint on where the new bins need to be added, and the average is probably enough to guide us.